File: scalanativelib/api/src/mill/scalanativelib/api/ScalaNativeWorkerApi.java
package mill.scalanativelib.api;

import sbt.testing.Framework;

public interface ScalaNativeWorkerApi {
    java.io.File discoverClang();
    java.io.File discoverClangPP();
    String[] discoverCompileOptions();
    String[] discoverLinkingOptions();

    NativeConfig config(String mainClass,
                        java.io.File[] classpath,
                        java.io.File nativeWorkdir,
                        java.io.File nativeClang,
                        java.io.File nativeClangPP,
                        java.util.Optional<String> nativeTarget,
                        String[] nativeCompileOptions,
                        String[] nativeLinkingOptions,
                        String nativeGC,
                        boolean nativeLinkStubs,
                        LTO nativeLTO,
                        ReleaseMode releaseMode,
                        boolean optimize,
                        boolean embedResources,
                        boolean incrementalCompilation,
                        NativeLogLevel logLevel);

    String defaultGarbageCollector();
    java.io.File nativeLink(NativeConfig nativeConfig, java.io.File outPath);

    GetFrameworkResult getFramework(java.io.File testBinary,
                                    java.util.Map<String, String> envVars,
                                    NativeLogLevel logLevel,
                                    String frameworkName);
}

File: scalanativelib/worker-api/src/mill/scalanativelib/worker/api/ScalaNativeWorkerApi.scala
package mill.scalanativelib.worker.api

import java.io.File

private[scalanativelib] trait ScalaNativeWorkerApi {

  def discoverClang(): File
  def discoverClangPP(): File
  def discoverCompileOptions(): Seq[String]
  def discoverLinkingOptions(): Seq[String]
  def defaultGarbageCollector(): String

  def config(
      mainClass: Either[String, String],
      classpath: Seq[File],
      nativeWorkdir: File,
      nativeClang: File,
      nativeClangPP: File,
      nativeTarget: Option[String],
      nativeCompileOptions: Seq[String],
      nativeLinkingOptions: Seq[String],
      nativeGC: String,
      nativeLinkStubs: Boolean,
      nativeLTO: String,
      releaseMode: String,
      nativeOptimize: Boolean,
      nativeEmbedResources: Boolean,
      nativeIncrementalCompilation: Boolean,
      nativeDump: Boolean,
      nativeMultithreading: Option[Boolean],
      nativeServiceProviders: Map[String, Seq[String]],
      logLevel: NativeLogLevel,
      buildTarget: BuildTarget
  ): Either[String, Object]

  def nativeLink(nativeConfig: Object, outPath: File): File

  def getFramework(
      testBinary: File,
      envVars: Map[String, String],
      logLevel: NativeLogLevel,
      frameworkName: String
  ): (() => Unit, sbt.testing.Framework)
}

private[scalanativelib] sealed abstract class NativeLogLevel(val value: Int)
private[scalanativelib] object NativeLogLevel {
  case object Error extends NativeLogLevel(200)
  case object Warn extends NativeLogLevel(300)
  case object Info extends NativeLogLevel(400)
  case object Debug extends NativeLogLevel(500)
  case object Trace extends NativeLogLevel(600)
}

private[scalanativelib] sealed trait BuildTarget
private[scalanativelib] object BuildTarget {
  case object Application extends BuildTarget
  case object LibraryDynamic extends BuildTarget
  case object LibraryStatic extends BuildTarget
}
